engine/scene/sceneloader.cpp

```diff
@@ -0,0 +1,198 @@
+        const std::string sceneName = sceneJson["name"];
+
+        ASSERT(sceneJson.contains("objects") && sceneJson["objects"].is_array());
+        for (const auto& gameObjectJson : sceneJson["objects"])
+        {
+            ASSERT(gameObjectJson.is_object());
+            ASSERT(gameObjectJson.contains("name") && gameObjectJson["name"].is_string());
+            ASSERT(gameObjectJson.contains("components") && gameObjectJson["components"].is_array());
+            for (const auto& componentJson : gameObjectJson["components"])
+            {
+                ASSERT(componentJson.is_object());
+                ASSERT(componentJson.contains("type") && componentJson["type"].is_string());
+                if (mComponentMappings.find(componentJson["type"]) == mComponentMappings.end())
+                {
+                    LogError("Unknown component type: {}", componentJson["type"]);
+                    ASSERT(false);
+                }
+            }
+        }
+
+        ASSERT(mScenes.find(sceneName) == mScenes.end());
+        mScenes[sceneName] = sceneJson;
+    }
+}
+
+std::vector<std::unique_ptr<GameObject>> SceneLoader::LoadScene(const std::string& sceneName)
+{
+    if (mScenes.find(sceneName) == mScenes.end())
+    {
+        LogError("Scene {} does not exist!", sceneName);
+        ASSERT(false);
+        return {};
+    }
+
+    const nlohmann::json& sceneJson = mScenes[sceneName];
+    return LoadSceneFromJson(sceneJson);
+}
+
+std::vector<std::unique_ptr<GameObject>> SceneLoader::LoadSceneFromJson(const nlohmann::json& sceneJson)
+{
+    std::vector<std::unique_ptr<GameObject>> gameObjects {};
+
+    const std::string sceneName = sceneJson["name"];
+    LogInfo("Loading Scene: {}", sceneName);
+
+    int totalComponents = 0;
+    const auto startTime = std::chrono::high_resolution_clock::now();
+
+    for (const auto& gameObjectJson : sceneJson["objects"])
+    {
+        const std::string gameObjectName = gameObjectJson["name"];
+
+        std::unique_ptr<GameObject> gameObject = std::make_unique<GameObject>();
+        gameObject->SetInstanceName(gameObjectName);
+
+        for (const auto& componentJson : gameObjectJson["components"])
+        {
+            const std::string componentTypeName = componentJson["type"];
+
+            // Add component to game object using the component mapping lambdas
+            BaseComponent* newComponent = mComponentMappings[componentTypeName](gameObject.get());
+            totalComponents++;
+
+            ComponentDeserializer componentDeserializer {};
+            componentDeserializer.DeserializeComponent(newComponent, componentJson);
+        }
+
+        gameObjects.push_back(std::move(gameObject));
+    }
+
+    const auto endTime = std::chrono::high_resolution_clock::now();
+    const auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(endTime - startTime).count();
+
+    LogInfo("Scene Loaded, Time Taken: {}ms ({}s). Total Objects: {} Total Components: {}",
```

Comment:
Cool!! I don't know if I missed this or forgot it over the holidays, but using `fmtlib` or something similar. I've really loved it when I've used it in the past.

No more remembering what the hell the format specifier for `unsigned size_t`